### PR TITLE
Add/checklist toggle affordance

### PR DESCRIPTION
**Pull request by fditrapani**
~**Note: This PR is still a work in progress. I have the visual styles applied but am waiting on another PR to merge to finalize the functionality. For more details checkout https://github.com/Automattic/zelda-private/issues/51**~

#### Changes proposed in this Pull Request

This PR:
- Adds an affordance to the checklist tasks so people know they can toggle them
- Synchronizes the Jetpack and WP.com checklists to both have the expand/collapse functionality

###### Screens
image

image

#### Testing instructions

In your browser console: `localStorage.setItem( 'debug', 'calypso:analytics:tracks' )`, and reload.

WordPress.com
* Create a new site
* Visit `/checklist/{your-url}` (First task should be open while others are closed. )
* Click on the task title to toggle the view.
* In the browser console, verify that expanding a task fires a `calypso_checklist_task_expand` Tracks event, with the correct `step_name` prop set.

Jetpack
* Create a new site with Jurrasic ninja
* Visit `/plans/my-plan/{your-url}` (First task should be open while others are closed. )
* Click on the task title to toggle the view.
* In the browser console, verify that expanding a task fires a `calypso_checklist_task_expand` Tracks event, with the correct `step_name` prop set.

WooCommerce
* Logged-in, navigate to http://calypso.localhost:3000/start/site-type
* Select 'Online Store' box and click continue.
* Go thru the usual NUX flow (select a domain, pick the eCommerce plan, and pay for your plan. You can skip adding G Suite).
* When presented with the purple 'Thank You' page, click the 'Create your Store' button.
* (The 'Building your Store' step can take a while.)
* Click 'Set up my Store'. Another 'Building your Store' step follows.
* When setting your address, it's important to enter a US or Canadian one.
* You're now presented with the Store checklist! From now on, you can always return to this site by navigating to `http://calypso.localhost:3000/store/<site>`
* Verify that the checklist works -- try toggling individual tasks

Fixes https://github.com/Automattic/zelda-private/issues/51

#### Follow-up Work

This is really looking a lot like a collection of [`FoldableCard`](http://calypso.localhost:3000/devdocs/design/foldable-card) components now, so it might make sense to actually use those to implement the checklist (their props interface permitting)

### CODE REVIEW

File: client/components/checklist/task.js [hunk 1]
```diff
@@ -178,45 +187,45 @@ class Task extends PureComponent {
...
 								borderless
 								className="checklist__task-title-link"
 								href={ href }
-								onClick={ this.onTaskClick }
+								onClick={ this.onTaskTitleClick }
 								target={ target }
 							>
-								{ title }
+								{ completed ? completedTitle : title }
+								{ this.renderChevron() }
```
Comment (jblz) [inline client/components/checklist/task.js]: This can be simplified:

```suggestion
								<Gridicon icon={ this.isTaskCollapsed() ? 'chevron-down' : 'chevron-up' } />;
```

...& we can remove the `renderChevron` function if we don't plan on expanding it straightaway.

Comment (fditrapani) [reply client/components/checklist/task.js]: That's much cleaner, thanks for the suggestion.

File: client/my-sites/plans/current-plan/jetpack-checklist/constants.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
+/**
```
Comment (sirreal) [inline client/my-sites/plans/current-plan/jetpack-checklist/constants.js]: This file may have been unintentionally added. Did you mean to edit `client/my-sites/plans/current-plan/jetpack-checklist/constants.ts`?

Comment (fditrapani) [reply client/my-sites/plans/current-plan/jetpack-checklist/constants.js]: oh yeah, that's weird. I don't know how that even happened. Thanks for catching that.

**Comment by jeffgolenski** [comment]
@fditrapani Unsure if related, but when I create a new WP site and head to the checklist I see it like this. The launch step is way out of place.

Screen Shot 2019-06-06 at 11 05 40 AM

**Comment by jeffgolenski** [comment]
Looking good on the jetpack side, albiet one thing - not sure if related to this PR. The verbiage for uncompleted tasks is incorrect. Currently it says "Change" - it should say "do it" or "try it"

Screen Shot 2019-06-06 at 11 13 59 AM
Screen Shot 2019-06-06 at 11 14 19 AM


Nice work filippo!

File: client/components/checklist/task.js [hunk 3]
```diff
@@ -143,9 +145,16 @@ class Task extends PureComponent {
 		return <Gridicon icon={ 'checkmark' } size={ 18 } />;
 	}
 
+	renderChevron = () => {
+		if ( this.isTaskCollapsed() ) {
+			return <Gridicon icon="chevron-down" className="checklist__toggle-affordance" />;
+		}
+
+		return <Gridicon icon="chevron-up" className="checklist__toggle-affordance" />;
```
Comment (ramonjd) [inline client/components/checklist/task.js]: Another way to do this might be to assign the class name based on the conditional check rather than printing out the component twice, e.g., 

```
renderChevron = () =>
    <Gridicon icon={ `chevron-${ this.isTaskCollapsed() ? 'down' : 'up' }` } className="checklist__toggle-affordance" />
```

File: client/components/checklist/task.js (hunk 1 above)
Comment (ramonjd) [reply client/components/checklist/task.js]: Ah, then ignore my comment https://github.com/Automattic/wp-calypso/pull/33666/files#r291427767 :)

**Comment by fditrapani** [comment]
Thanks for the feedback @jeffgolenski 

> Unsure if related, but when I create a new WP site and head to the checklist I see it like this. The launch step is way out of place.

This is related to some ongoing checklist work for the WP.com. It only shows up in our dev environments. 

> The verbiage for uncompleted tasks is incorrect. Currently it says "Change" - it should say "do it" or "try it"

I haven't made any changes to the copy. We can address that in a separate PR.

File: client/my-sites/plans/current-plan/jetpack-checklist/index.tsx [hunk 4]
```diff
@@ -241,6 +296,7 @@ class JetpackChecklist extends PureComponent< Props > {
 									? `/media/${ siteSlug }`
 									: `/settings/performance/${ siteSlug }`
 							}
+							id={ 'video-hosting' }
```
Comment (ockham) [inline client/my-sites/plans/current-plan/jetpack-checklist/index.tsx]: Nitpick, for string literal props in JSX, it's preferable to use 'HTML-style' attribute notation:

```suggestion
							id="video-hosting"
```

(I'll change while rebasing)

File: client/components/checklist/style.scss [hunk 5]
```diff
@@ -11,7 +15,7 @@
 		padding: 0;
 	}
 
-	&-main {
+	.checklist__header-main {
```
Comment (ockham) [inline client/components/checklist/style.scss]: Curious, why the change to spelling out the classname explicitly? To make it easier to `grep`?

File: client/components/checklist/task.js [hunk 6]
```diff
@@ -106,25 +105,27 @@ class Task extends PureComponent {
...
 
-		// We presuppose expandability based on the presence of `onCollapsedClick`
-		if ( ! completed && this.isTaskCollapsed() && onCollapsedClick ) {
-			event.preventDefault();
-			onCollapsedClick( id );
-		} else {
-			// The task-related action
-			onClick();
-		}
+		this.isTaskCollapsed() ? this.props.onTaskTitleClick( id ) : onTaskTitleClick( 'none' );
```
Comment (ockham) [inline client/components/checklist/task.js]: Let's use either `this.props.onTaskTitleClick` _or_ `onTaskTitleClick` consistently:

```suggestion
		this.isTaskCollapsed() ? onTaskTitleClick( id ) : onTaskTitleClick( 'none' );
```

File: client/components/checklist/style.scss (hunk 5 above)
Comment (simison) [reply client/components/checklist/style.scss]: I also prefer full explicit class names for clarity (just how my brain works) as well for easier searching. No strong opinion the way or the other tho. :-)

Comment (fditrapani) [reply client/components/checklist/style.scss]: Yeah, I did it for searching plus it's also easier to read.

File: client/my-sites/plans/current-plan/jetpack-checklist/index.tsx (hunk 4 above)
Comment (fditrapani) [reply client/my-sites/plans/current-plan/jetpack-checklist/index.tsx]: > (I'll change while rebasing)

Noted and thank you for picking this up. I'll be travelling for the design meetup today but let me know if you have any questions or when would be a good time to check it out again.

File: client/components/checklist/task.js (hunk 6 above)
Comment (fditrapani) [reply client/components/checklist/task.js]: > Let's use either this.props.onTaskTitleClick or onTaskTitleClick consistently:

Can you elaborate on this a little more? Is it in this piece of code or elsewhere?

File: client/my-sites/plans/current-plan/jetpack-checklist/index.tsx [hunk 7]
```diff
@@ -115,18 +119,45 @@ class JetpackChecklist extends PureComponent< Props > {
...
 
+	onTaskTitleClick = selectedTaskId => {
+		this.setState( { selectedTaskId } );
+		this.props.recordTracksEvent( 'calypso_checklist_task_expand', {
+			step_name: selectedTaskId,
+			product: 'Jetpack',
+		} );
+	};
+
+	getFirstIncompleteTask = taskStatuses => {
```
Comment (fditrapani) [inline client/my-sites/plans/current-plan/jetpack-checklist/index.tsx]: This is some duplicated code in both the Wp.com checklist and Jetpack code. I'm not sure where to put it or how to code it in such a way that it becomes centralized. Any thoughts @ockham or @simison?

File: client/components/checklist/style.scss [hunk 8]
```diff
@@ -138,19 +132,21 @@
 		bottom: 0;
 		left: 34px;
 		border-left: 1px solid var( --color-neutral-100 );
+		z-index: 1;
```
Comment (ockham) [inline client/components/checklist/style.scss]: Should we avoid hard-coded z-indices in favor of using the logic from [this file](https://github.com/Automattic/wp-calypso/blob/3b3dc15cf9818f469b9efa72c410248d7d7084eb/assets/stylesheets/shared/functions/_z-index.scss)?

File: client/components/checklist/style.scss [hunk 9]
```diff
@@ -138,19 +132,21 @@
...
-		top: 18px;
+		top: 16px;
 		left: 24px;
 		width: 16px;
 		height: 16px;
 		border: 2px solid var( --color-neutral-100 );
 		border-radius: 16px;
 		background: var( --color-white );
 		cursor: pointer;
+		z-index: 2;
```
Comment (ockham) [inline client/components/checklist/style.scss]: Same as https://github.com/Automattic/wp-calypso/pull/33666/files#r292497101

File: client/components/checklist/style.scss [hunk 10]
```diff
@@ -188,83 +184,79 @@
...
+		text-align: left;
+		color: var( --color-neutral-dark );
+		font-weight: 600;
+		font-size: 16px;
+		border-radius: 0;
 
-		@include breakpoint( '<480px' ) {
-			padding-left: 0;
-			align-items: center;
+		.checklist__toggle-affordance {
```
Comment (ockham) [inline client/components/checklist/style.scss]: I'm personally a bit confused by the term 'affordance' :sweat_smile: Is that a standard UI design term, without any good alternatives? Are there maybe some synonyms that are clearer to us developer folk? :smile:

File: client/components/checklist/task.js (hunk 6 above)
Comment (ockham) [reply client/components/checklist/task.js]: Yes: of the two logical 'branches' of the ternary, one calls `this.props.onTaskTitleClick()`, while the other calls `onTaskTitleClick()`. However, both are the same thing -- the latter is being defined (via destructuring) on [R110](https://github.com/Automattic/wp-calypso/pull/33666/files#diff-acb49564d5ae2fcd2a24d573341a2d3dR110).

Thus, for consistency's sake, we should decide on one of them and use it in both cases. (Probably `onTaskTitleClick` since it wouldn't make much sense to destructure it otherwise.)

File: client/components/checklist/style.scss [hunk 11]
```diff
@@ -188,83 +184,79 @@
...
+		font-size: 14px;
+		color: var( --color-neutral-dark );
+	}
 
-			.checklist__task-secondary & {
-				display: inline-block;
-				margin-left: 1.5em;
-			}
-		}
+	.checklist__task-action-wrapper {
```
Comment (ockham) [inline client/components/checklist/style.scss]: Do we need the extra wrapper `<div />`? Can't we just assign the bottom margin to `.checklist__task-action`?

File: client/my-sites/plans/current-plan/jetpack-checklist/constants.ts [hunk 12]
```diff
@@ -100,12 +100,14 @@ export const JETPACK_CHECKLIST_TASK_AKISMET: TaskUiDescription = {
 	title: translate( "We're automatically turning on spam filtering." ),
 	completedButtonText: translate( 'View spam stats' ),
 	completedTitle: translate( "We've automatically turned on spam filtering." ),
+	tourId: 'jetpackAkismet',
```
Comment (ockham) [inline client/my-sites/plans/current-plan/jetpack-checklist/constants.ts]: Why are we adding this? There is no such tour, and the task is being marked as automatically completed.

File: client/my-sites/checklist/wpcom-checklist/component.jsx [hunk 13]
```diff
@@ -162,10 +162,11 @@ class WpcomChecklistComponent extends PureComponent {
 		} );
 	};
 
-	onCollapsedClick = selectedTaskId => {
+	onTaskTitleClick = selectedTaskId => {
```
Comment (ockham) [inline client/my-sites/checklist/wpcom-checklist/component.jsx]: This isn't only called when clicking the task's title, but also the chevron. Should we thus rename it to `onTaskClick`?

(We tried to implement a convention to actually make these action handler function names descriptive wrt what they do as opposed to when they're called -- in this case, we could call it e.g. `expandTask`.)

Comment (ockham) [reply client/my-sites/checklist/wpcom-checklist/component.jsx]: Also, this is generic enough to the checklist UI component that it should probably be moved into `components/checklist` (and removed here and from the JP checklist).

File: client/components/checklist/task.js (hunk 6 above)
Comment (ockham) [reply client/components/checklist/task.js]: Oh, I just realized that the method that we're calling this from has the exact same name :thinking: Okay, that could lead to some confusion.

Renaming the method to `expandTask` could provide an easy way out (see https://github.com/Automattic/wp-calypso/pull/33666/files#r292546081)

Comment (ockham) [reply client/components/checklist/task.js]: Finally, I don't think we should use `'none'` as the magical value to indicate that the task is collapsed. We're already using `null` for the `initialState` [here](https://github.com/Automattic/wp-calypso/pull/33666/files#diff-109d16ece1aaa451541aafc75a84baf4R71).

File: client/my-sites/plans/current-plan/jetpack-checklist/index.tsx [hunk 14]
```diff
@@ -115,18 +119,45 @@ class JetpackChecklist extends PureComponent< Props > {
...
+					id={ taskId }
+					key={ taskId }
 					onClick={ this.handleTaskStart( {
 						taskId,
 						tourId: get( task, [ 'tourId' ] ),
 					} ) }
+					onTaskTitleClick={ this.onTaskTitleClick }
 					title={ task.title }
-					key={ taskId }
+					selectedTaskId={ this.state.selectedTaskId }
```
Comment (ockham) [inline client/my-sites/plans/current-plan/jetpack-checklist/index.tsx]: From an API design POV, a component named `Task` -- used beside many other `Task` components as children of a `Checklist` -- shouldn't have a prop named `selectedTaskId`. The `Task` component doesn't care about task IDs -- all it cares about is whether it's selected or not.

A better choice is thus a `selected` (or maybe `expanded`) boolean prop.

File: client/my-sites/plans/current-plan/jetpack-checklist/index.tsx [hunk 15]
```diff
@@ -115,18 +119,45 @@ class JetpackChecklist extends PureComponent< Props > {
 					completedTitle={ task.completedTitle }
 					description={ task.description }
 					duration={ task.duration }
+					firstIncomplete={ this.getFirstIncompleteTask( taskIds ) }
```
Comment (ockham) [inline client/my-sites/plans/current-plan/jetpack-checklist/index.tsx]: Similar as https://github.com/Automattic/wp-calypso/pull/33666/files#r292552161. The `Task` component shouldn't need to care what's the first incomplete task, that's the `Checklist'`s job.

File: client/my-sites/plans/current-plan/jetpack-checklist/index.tsx (hunk 7 above)
Comment (ockham) [reply client/my-sites/plans/current-plan/jetpack-checklist/index.tsx]: Good thinking! If it's generically applicable to checklist, it should be moved to `components/checklist/`.

File: client/my-sites/plans/current-plan/jetpack-checklist/index.tsx [hunk 16]
```diff
@@ -115,18 +119,45 @@ class JetpackChecklist extends PureComponent< Props > {
...
+		Object.values( taskStatuses ).find( function( item, index ) {
+			if ( item.completed === false ) {
+				taskObject.id = Object.keys( taskStatuses )[ index ];
+				taskObject.isCompleted = item.completed;
+			}
+
+			return item.completed === false;
+		} );
+
+		return taskObject;
```
Comment (ockham) [inline client/my-sites/plans/current-plan/jetpack-checklist/index.tsx]: The `getFirstIncompleteTask` logic could probably be shortened to something like https://github.com/Automattic/wp-calypso/blob/60c089ce54589232d998a0d2f159c7536db8bc40/client/my-sites/checklist/wpcom-checklist/wpcom-task-list.js#L125-L127

Also, it probably makes sense moving that logic into `components/checklist`.

File: client/my-sites/plans/current-plan/jetpack-checklist/index.tsx [hunk 17]
```diff
@@ -162,15 +199,23 @@ class JetpackChecklist extends PureComponent< Props > {
 				>
 					<Task
 						{ ...JETPACK_CHECKLIST_TASK_PROTECT }
+						id="jetpack_task_protect"
+						onTaskTitleClick={ this.onTaskTitleClick }
+						selectedTaskId={ this.state.selectedTaskId }
```
Comment (ockham) [inline client/my-sites/plans/current-plan/jetpack-checklist/index.tsx]: `Task` shouldn't have to know about `selectedTaskId`, see https://github.com/Automattic/wp-calypso/pull/33666/files#r292552161

File: client/my-sites/plans/current-plan/jetpack-checklist/index.tsx [hunk 18]
```diff
@@ -162,15 +199,23 @@ class JetpackChecklist extends PureComponent< Props > {
 				>
 					<Task
 						{ ...JETPACK_CHECKLIST_TASK_PROTECT }
+						id="jetpack_task_protect"
+						onTaskTitleClick={ this.onTaskTitleClick }
```
Comment (ockham) [inline client/my-sites/plans/current-plan/jetpack-checklist/index.tsx]: The fact that we're adding the same `onTaskTitleClick` method to all `Task` components indicates that it should be absorbed into the generic base component, `components/checklist`.

File: client/my-sites/plans/current-plan/jetpack-checklist/index.tsx [hunk 19]
```diff
@@ -180,25 +225,34 @@ class JetpackChecklist extends PureComponent< Props > {
...
 							href={ JETPACK_CHECKLIST_TASK_AKISMET.getUrl( siteSlug ) }
+							firstIncomplete={ firstIncomplete }
 							inProgress={ ! akismetFinished }
 							onClick={ this.handleTaskStart( { taskId: 'jetpack_spam_filtering' } ) }
 							target="_blank"
 						/>
 					) }
-					{ this.renderTaskSet( JETPACK_SECURITY_CHECKLIST_TASKS ) }
+
+					{ this.renderTaskSet( JETPACK_SECURITY_CHECKLIST_TASKS, firstIncomplete ) }
```
Comment (ockham) [inline client/my-sites/plans/current-plan/jetpack-checklist/index.tsx]: Maybe I messed this up when rebasing, but [`renderTaskSet`](https://github.com/Automattic/wp-calypso/pull/33666/files#diff-109d16ece1aaa451541aafc75a84baf4R99) currently doesn't seem to accept a second arg.

**Comment by ockham** [comment]
This is looking quite good! Hope the number of my comments isn't too discouraging, most are minor.

I've noticed that this PR is removing `buttonPrimary` functionality, but this seems still used elsewhere (try `grep`ping), so we should remove those remaining instances.

Most importantly, it's used by the WooCommerce checklist -- we should definitely test that checklist before shipping these changes :slightly_smiling_face:

**Comment by ockham** [comment]
TODO:
- Implement expand-first-incomplete-task logic
- Fix WP.com checklist
- Fix Woo Checklist

File: client/my-sites/plans/current-plan/jetpack-checklist/index.tsx [hunk 20]
```diff
@@ -158,17 +165,20 @@ class JetpackChecklist extends PureComponent< Props > {
 				<Checklist
 					className="jetpack-checklist"
 					isPlaceholder={ ! taskStatuses }
+					onExpandTask={ this.trackExpandTask }
 					progressText={ translate( 'Your Jetpack setup progress' ) }
 				>
 					<Task
 						{ ...JETPACK_CHECKLIST_TASK_PROTECT }
+						id="jetpack_task_protect"
```
Comment (ockham) [inline client/my-sites/plans/current-plan/jetpack-checklist/index.tsx]: I tried moving these `id`s to `./constants.ts`, but that's anything but trivial, due to the vastly heterogenous handling of task IDs: e.g. for some IDs, we use object keys in a collection of tasks -- but not all tasks are contained in a collection. Since the constants file is typed, we'd have to add an `id: string?` field, i.e. a conditional ID, which would be _very_ atypical for an ID. (In other words, typing the constants file reveals that there's too much heterogenity in how we handle our tasks.)

My conclusion is thus that we should probably either move all task definitions from this file to `./constants.ts`, or vice versa (merge `./constants.ts` into this file). Either way, that's for another PR.

Comment (sirreal) [reply client/my-sites/plans/current-plan/jetpack-checklist/index.tsx]: I don't understand why the tasks live spread across two files either. I do think it's OK to add the option `id`. Types should be serving us by providing better insights and safety. We don't want them to strictly get in the way.

This may be types highlighting a situation that is hard to justify. That's relevant and something we should step back and handle as you suggest. That said, I don't think the fact that types exist should prevent us from describing an existing sub-optimal situation. They can continue to help us describe the situation as it is now. In the future when we come to an improved longer-term solution they'll help us to clean up as well 🙂 

```patch
diff --git a/client/my-sites/plans/current-plan/jetpack-checklist/constants.ts b/client/my-sites/plans/current-plan/jetpack-checklist/constants.ts
index 1c19143853..954d7a6ed0 100644
--- a/client/my-sites/plans/current-plan/jetpack-checklist/constants.ts
+++ b/client/my-sites/plans/current-plan/jetpack-checklist/constants.ts
@@ -19,6 +19,7 @@ export function getJetpackChecklistTaskDuration( minutes: number ): string {
 }
 
 interface TaskUiDescription {
+	readonly id?: string;
 	readonly title: string;
 	readonly description?: string;
 	readonly completedButtonText: string;
@@ -105,6 +106,7 @@ export const JETPACK_CHECKLIST_TASK_AKISMET: TaskUiDescription = {
 };
 
 export const JETPACK_CHECKLIST_TASK_PROTECT: TaskUiDescription = {
+	id: 'jetpack_task_protect',
 	title: translate( "We're automatically protecting you from brute force login attacks." ),
 	completedTitle: translate( "We've automatically protected you from brute force login attacks." ),
 	completedButtonText: translate( 'Configure' ),
diff --git a/client/my-sites/plans/current-plan/jetpack-checklist/index.tsx b/client/my-sites/plans/current-plan/jetpack-checklist/index.tsx
index cb0d1dfdf1..57d261db29 100644
--- a/client/my-sites/plans/current-plan/jetpack-checklist/index.tsx
+++ b/client/my-sites/plans/current-plan/jetpack-checklist/index.tsx
@@ -170,7 +170,6 @@ class JetpackChecklist extends PureComponent< Props > {
 				>
 					<Task
 						{ ...JETPACK_CHECKLIST_TASK_PROTECT }
-						id="jetpack_task_protect"
 						completed
 						href={ JETPACK_CHECKLIST_TASK_PROTECT.getUrl( siteSlug ) }
 						onClick={ this.handleTaskStart( { taskId: 'jetpack_protect' } ) }
```

**Comment by ockham** [comment]
Putting this on hold for the following reason:

The removal of the `completedDescription` prop has actually a big (UI) impact for one WP.com task (email forwarding):

image

That task is only shown under certain circumstances (see #29616 for a matrix), but whenever it is shown, it is shown as complete. _This_ is why there is a `completedDescription` prop, and to make things more complicated, it requires a link that kinda serves as a CTA.

Retaining a `completedDescription` (to be shown when a task is collapsed?) is arguably against the UI pattern of this PR, and it's doubtful if it makes sense to go out of our way to account for the one task off that is arguably a bit of an antipattern anyway:

One could argue that it's a step to be completed, so it should be made an uncompleted task. However, since it involves upgrading the user's plan, that would mean that the user can only complete 100% of the checklist if they spend more money -- when the checklist is really meant to help them get acquainted with the plan they bought it the first place.

**Comment by ockham** [comment]
The 'regular' email step is a problem though:

image

This happens because the `Task` component is wrapped in a `Fragment` (so it also renders some notice component), which in turn means the `React.Children.map()`/`React.cloneElement()` approach I chose for expand/collapse handling isn't working.

https://github.com/Automattic/wp-calypso/blob/c2a8a12e0beffdef70fe8d341b90d7bf16b6511a/client/my-sites/checklist/wpcom-checklist/component.jsx#L704-L751

**Comment by ockham** [comment]
Okay, I think this should be ready for review at last :crossed_fingers:

File: client/components/checklist/task.js [hunk 21]
```diff
@@ -172,50 +154,50 @@ class Task extends PureComponent {
...
-					{ duration && (
-						<small className="checklist__task-duration">
-							{ translate( 'Estimated time:' ) } { duration }
-						</small>
+
+					{ ! collapsed && (
+						<div className="checklist__task-content">
+							<p className="checklist__task-description">{ description }</p>
+
+							<div className="checklist__task-action-duration-wrapper">
```
Comment (simison) [inline client/components/checklist/task.js]: Nit; this class could be `-actions` or something like that, since it wraps duration _and_ action, as well as possible future "skip" button.

Not a blocker.

**Comment by ockham** [comment]
> Observations about focus color:
> 
> * On mobile, when toggling the same task open/closed, the focus color stays active after closing the task: https://cloudup.com/cy-xzL3SAuz
> * It's a bit confusing how the little line between checkbox disappears on focus. I expected it to stay visible and just change color little bit:
>   Screenshot 2019-06-17 at 11 18 17
>   Screenshot 2019-06-17 at 11 18 03
> 
> Not a blocker and something we can loop back after merging if we decide to change it.

Ah, maybe this is what the `z-index` was for (which I removed per 61d969e).

**Comment by ockham** [comment]
> Found an issue — when these spinners are spinning, we should hide "do it" buttons for those tasks:

Should be fixed by 97f5b20. (How do you test that? Those tasks have already been finished when I reach the page...)

File: client/my-sites/plans/current-plan/jetpack-checklist/constants.ts [hunk 22]
```diff
@@ -105,12 +107,15 @@ export const JETPACK_CHECKLIST_TASK_AKISMET: TaskUiDescription = {
 };
 
 export const JETPACK_CHECKLIST_TASK_PROTECT: TaskUiDescription = {
-	title: translate( "We've automatically protected you from brute force login attacks." ),
+	id: 'jetpack_task_protect',
+	title: translate( "We're automatically protecting you from brute force login attacks." ),
```
Comment (a8ci18n) [inline client/my-sites/plans/current-plan/jetpack-checklist/constants.ts]: :ok: This change is safe, we'll keep the existing 20 translations.

**Comment by sirreal** [comment]
> How do you test that? Those tasks have already been finished when I reach the page...

@ockham Should be testable with a new Jetpack site or by deactivating/removing VaultPress and/or Akismet.

**Comment by ockham** [comment]
> > How do you test that? Those tasks have already been finished when I reach the page...
> 
> @ockham Should be testable with a new Jetpack site or by deactivating/removing VaultPress and/or Akismet.

Oh, I was actually trying with a new jurassic.ninja site, should've mentioned :sweat_smile:

**Comment by ockham** [comment]
> > Observations about focus color:
> > 
> > * On mobile, when toggling the same task open/closed, the focus color stays active after closing the task: https://cloudup.com/cy-xzL3SAuz
> > * It's a bit confusing how the little line between checkbox disappears on focus. I expected it to stay visible and just change color little bit:
> >   Screenshot 2019-06-17 at 11 18 17
> >   Screenshot 2019-06-17 at 11 18 03
> > 
> > Not a blocker and something we can loop back after merging if we decide to change it.
> 
> Ah, maybe this is what the `z-index` was for (which I removed per [61d969e](https://github.com/Automattic/wp-calypso/commit/61d969e94138551c5f728b9d3f8d3d2707a6992e)).

Yep, seems to work if I revert. However, I'd rather not hard-code z-indices here (https://github.com/Automattic/wp-calypso/pull/33666#discussion_r292497101). But yeah, this is minor enough that we shouldn't block this PR by it.

Maybe a designer can help us with this in a follow-up PR. (Note that we're also currently using both a `checklist__task` and `checklist__task_wrapper` here, so there should be plenty of oppurtunity to implement the same visual effect through various alternatives.)

**Comment by simison** [comment]
> Found an issue — when these spinners are spinning, we should hide "do it" buttons for those tasks: [...]
>
> Should be fixed by 97f5b20. (How do you test that? Those tasks have already been finished when I reach the page...)

There are (sadly) a few ways to get the site to that broken state but one of the easiest is to pick expired Jurassic Ninja site from the site picker while at the checklist. :-) Another is to add `define( 'DISALLOW_FILE_MODS', true );` in your site's config.

Now I see this — no toggles but "do it" button is visible:

image

**Comment by sirreal** [comment]
I see duplicated tasks (completed and in progress) as well as "Do It!" buttons for plan provisioning steps (shouldn't have a Do It button at all):

Screen Shot 2019-06-17 at 15 52 49

To reproduce:
- Set up a Jetpack site with a plan
- After initial setup, visit WP Admin and deactivate and remove VaultPress and Akismet
- Return the the checklist (`/plans/my-plan/SITE_SLUG`)

File: client/my-sites/plans/current-plan/jetpack-checklist/constants.ts (hunk 22 above)
Comment (a8ci18n) [inline client/my-sites/plans/current-plan/jetpack-checklist/constants.ts]: :ok: This change is safe, we'll keep the existing 20 translations.

**Comment by ockham** [comment]
Thanks @sirreal and @simison, the bug was in the methdod that falls back to expanding the first uncompleted task. Added logic to also make sure it's not `inProgress`: 147079e

**Comment by sirreal** [comment]
As tasks load, we may lose our position:

jump

This seems minor to me, I wouldn't be opposed to merging and opening a known issue.

File: client/components/checklist/style.scss (hunk 10 above)
Comment (fditrapani) [reply client/components/checklist/style.scss]: yes, it is a standard UI term. We could change it to `checklist__toggle-icon` — that's probably clearer.

File: client/components/checklist/style.scss (hunk 11 above)
Comment (fditrapani) [reply client/components/checklist/style.scss]: The wrapper `divs` are here to help with hiding/showing the details and also for some visual styling.

File: client/components/checklist/style.scss (hunk 10 above)
Comment (simison) [reply client/components/checklist/style.scss]: I think that would be clearer yeah 👍

Comment (fditrapani) [reply client/components/checklist/style.scss]: Updated!

File: client/my-sites/plans/current-plan/jetpack-checklist/constants.ts (hunk 22 above)
Comment (a8ci18n) [inline client/my-sites/plans/current-plan/jetpack-checklist/constants.ts]: :ok: This change is safe, we'll keep the existing 20 translations.

Comment (a8ci18n) [inline client/my-sites/plans/current-plan/jetpack-checklist/constants.ts]: :ok: This change is safe, we'll keep the existing 20 translations.
